packages/angular/src/migrations/update-16-1-0/extract-standalone-config-from-bootstrap.ts

```diff
@@ -0,0 +1,187 @@
+}
+
+function getImportTokenMap(bootstrapCallFileContentsAst: SourceFile) {
+  const importTokenMap = new Map<string, string>();
+  const importedTokensNodes = tsquery(
+    bootstrapCallFileContentsAst,
+    'ImportDeclaration > ImportClause',
+    { visitAllChildren: true }
+  );
+
+  for (const node of importedTokensNodes) {
+    importTokenMap.set(node.getText(), node.parent.getText());
+  }
+  return importTokenMap;
+}
+
+function getImportsRequiredForAppConfig(
+  importTokenMap: Map<string, string>,
+  appConfigNode: Node
+) {
+  const importsRequiredForAppConfig = new Set<string>();
+  const checkImportsForTokens = (nodeText: string) => {
+    const keys = importTokenMap.keys();
+    for (const key of keys) {
+      if (key.includes(nodeText))
+        importsRequiredForAppConfig.add(importTokenMap.get(key));
+    }
+  };
+  const visitEachChild = (node: Node) => {
+    node.forEachChild((node) => {
+      const nodeText = node.getText();
+      checkImportsForTokens(nodeText);
+      visitEachChild(node);
+    });
+  };
+  visitEachChild(appConfigNode);
+  return importsRequiredForAppConfig;
+}
+
+function getAppConfigFileContents(
+  importsRequiredForAppConfig: Set<string>,
+  appConfigText: string
+) {
+  const buildAppConfigFileContents = (
+    importStatements: string[],
+    appConfig: string
+  ) => `import { ApplicationConfig } from '@angular/core';${importStatements.join(
+    '\n'
+  )}
+        export const appConfig: ApplicationConfig = ${appConfig}`;
+
+  const appConfigFileContents = buildAppConfigFileContents(
+    Array.from(importsRequiredForAppConfig),
+    appConfigText
+  );
+  return appConfigFileContents;
+}
+
+function getBootstrapCallFileContents(
+  bootstrapCallFileContents: string,
+  appConfigNode: Node,
+  importsRequiredForAppConfig: Set<string>
+) {
+  let newBootstrapCallFileContents = `import { appConfig } from './app/app.config';
+        ${bootstrapCallFileContents.slice(
+          0,
+          appConfigNode.getStart()
+        )}appConfig${bootstrapCallFileContents.slice(appConfigNode.getEnd())}`;
+
+  for (const importStatement of importsRequiredForAppConfig) {
+    newBootstrapCallFileContents = newBootstrapCallFileContents
+      .split(importStatement)
+      .join('');
+  }
+  return newBootstrapCallFileContents;
+}
+
+export default async function extractStandaloneConfig(tree: Tree) {
+  if (!tsModule) {
+    tsModule = ensureTypescript();
+  }
+  if (!tsquery) {
+    tsquery = require('@phenomnomnominal/tsquery').tsquery;
+  }
+
+  const projects = getProjects(tree);
+
+  const BOOTSTRAP_APPLICATION_CALL_SELECTOR =
+    'CallExpression:has(Identifier[name=bootstrapApplication])';
+  const BOOTSTRAP_APPLICATION_CALL_CONFIG_SELECTOR =
+    'CallExpression:has(Identifier[name=bootstrapApplication]) > ObjectLiteralExpression';
+
+  for (const [projectName, project] of projects.entries()) {
+    if (project.projectType === 'application') {
+      if (project.targets?.build?.options?.main) {
```

Comment:
nit: we could reduce nesting
```suggestion
    if (project.projectType !== 'application' || !project.targets?.build?.options?.main) {
      continue;
    }
```